WIP - Argocd/Flux Comparison Document (#47)
open-gitops/documents: LANDSCAPE/COMPARISONS/ARGOCD-FLUX/index.md

> # Continuous Deployment on Kubernetes: What’s the Difference Between Argo CD and FluxCD

Reviewer: ```suggestion
# Continuous Deployment on Kubernetes: What’s the Difference Between Argo CD and Flux CD
```

Reviewer 2: Flux is just called "Flux".

Reviewer: Thanks @squaremo

Reviewer 2: Thank you for following through on my rather terse comment :pray:

(Context: Flux's github org and website domain use "fluxcd" ("flux" was gone already), which has caused ongoing confusion, even amongst its maintainers...)

> If you’ve heard of GitOps and its benefits, you may have come across two open source projects in the Cloud Native Computing Foundation: Argo CD and FluxCD. Members of these projects are involved in the CNCF’s OpenGitOps initiative, and both are incubating projects. With GitOps, you increase reliability and velocity, lower downtime, and can enjoy other benefits through key GitOps capabilities such as reconciliation.

Reviewer: ```suggestion
If you’ve heard of GitOps and its benefits, you may have come across two open source projects in the Cloud Native Computing Foundation: Argo CD and Flux CD. Members of these projects are involved in the CNCF’s OpenGitOps initiative, and both are incubating projects. With GitOps, you increase reliability and velocity, lower downtime, and can enjoy other benefits through key GitOps capabilities such as reconciliation.
```

Reviewer: ```suggestion
If you’ve heard of GitOps and its benefits, you may have come across two open source projects in the Cloud Native Computing Foundation: Argo CD and Flux CD. Members of these projects are involved in the CNCF’s OpenGitOps initiative, and both are incubating projects. With GitOps, you increase reliability and velocity, lower downtime, and can enjoy other benefits through key GitOps capabilities such as reconciliation.
```

> When building out a Kubernetes cluster, the actual pipelines for the code behind applications is just as important as the firewalls, load balancers and provisioning systems at the front. There are many ways to build out application CI/CD pipelines in Kubernetes, but in this article, we are going to focus very specifically on two options for Continuous Deployment: [Flux CD](https://fluxcd.io/) (specifically Flux v2) and [Argo CD](https://argoproj.github.io/cd/).

Reviewer: CI/CD → CI & CD
```suggestion
When building out a Kubernetes cluster, the actual pipelines for the code behind applications is just as important as the firewalls, load balancers and provisioning systems at the front. There are many ways to build out application CI & CD pipelines in Kubernetes, but in this article, we are going to focus very specifically on two options for Continuous Deployment: [Flux CD](https://fluxcd.io/) (specifically Flux v2) and [Argo CD](https://argoproj.github.io/cd/).
```

Reviewer 3: The most common reference is CI/CD not CI & CD.

Reviewer: @chris-short Understood.

How about removing the phrase "CI/CD"?

application CI/CD pipelines → application pipelines

```suggestion
When building out a Kubernetes cluster, the actual pipelines for the code behind applications is just as important as the firewalls, load balancers and provisioning systems at the front. There are many ways to build out application pipelines in Kubernetes, but in this article, we are going to focus very specifically on two options for Continuous Deployment: [Flux CD](https://fluxcd.io/) (specifically Flux v2) and [Argo CD](https://argoproj.github.io/cd/).
```
‎
For reference: [AWS Container Day: Kubernetes Edition KubeCon 2020](https://www.youtube.com/watch?v=o2_sGFkYNvA&t=1184s) by @countspongebob
> It is really critical to not conflate CI and CD. I try hard to not use the phrase CI/CD because they are very different practices. One of the lovely things about the GitOps model is the clean and understandable relationship that is created between CI as a Dev practice, and CD as an Ops practice. I think the notion of the Immutability Firewall in between them based on Git is really excellent.

Below are the results of exploring the differences between these two projects. Both have their uses, and both are quite well maintained and have active communities. This article will focus on those areas where the two differ, rather than on the overall CD landscape in which they exist.

> This article was written with the guidance of the [GitOps Working Group](https://github.com/gitops-working-group/gitops-working-group).

Reviewer: gitops-working-group/gitops-working-group → cncf/tag-app-delivery/tree/main/gitops-wg
```suggestion
This article was written with the guidance of the [GitOps Working Group](https://github.com/cncf/tag-app-delivery/tree/main/gitops-wg).
```

   </td>
  </tr>
  <tr>
   <td>Notes:
   </td>
   <td>
<ul>

<li>Flux CD is in the process of graduation from CNCF

<li>Flux has recently <a href="https://fluxcd.io/blog/2021-11-10-flux-security-audit/">completed a security audit </a>commissioned by the CNCF. They are expected to graduate very soon.
</li>
</ul>
   </td>
   <td>
<ul>

<li>Argo CD currently holds <a href="https://thenewstack.io/argocd-gets-devsecops-nod-with-cncf/">“adopt” status in DevSecOps from the CNCF</a> 

<li>Argo CD is expected to graduate in EOCY 2022

<li>Argo CD has undergone <a href="https://github.com/argoproj/argoproj/blob/master/docs/argo_security_final_report.pdf">an independent security audit</a> to comply with the latest security standards
</li>
</ul>
   </td>
  </tr>
</table>

# Installation/Bootstrapping

Both Flux and Argo need to be installed on the cluster you want to manage. Once installed, they need to be configured to read YAML files from a Git repository. This initial configuration step is called "bootstrapping". Flux and Argo have different paradigms to installing/bootstrapping.

## Flux CD

> FluxCD [installation/bootstrapping](https://fluxcd.io/docs/cmd/flux_bootstrap/) mechanism can be done together. The prerequisites for bootstrapping can be found in the [upstream documentation](https://fluxcd.io/docs/get-started/). During bootstrapping, Flux will create a repository on GitHub or GitLab (A generic “git” mode can also be used) and then create all the manifests for installation and connection to the git repo in that same git repo. The Flux installer will then “hand off” the responsibility of managing/syncing the cluster to the_ flux-system_ that is installed on the cluster, pointing to the repo that was just created.

Reviewer: ```suggestion
Flux CD [installation/bootstrapping](https://fluxcd.io/docs/cmd/flux_bootstrap/) mechanism can be done together. The prerequisites for bootstrapping can be found in the [upstream documentation](https://fluxcd.io/docs/get-started/). During bootstrapping, Flux will create a repository on GitHub or GitLab (A generic “git” mode can also be used) and then create all the manifests for installation and connection to the git repo in that same git repo. The Flux installer will then “hand off” the responsibility of managing/syncing the cluster to the_ flux-system_ that is installed on the cluster, pointing to the repo that was just created.
```

Flux can also be installed by [just using manifests](https://fluxcd.io/docs/installation/#dev-install), for those that don’t want to use bootstrapping.

## Argo CD

Argo CD has no native mechanism for bootstrapping. The [install](https://argo-cd.readthedocs.io/en/stable/getting_started/#1-install-argo-cd) is just that: the install of the Argo CD components. It leaves the rest of the configuration up to the end user. The install is straightforward. It is done via a YAML file using the `kubectl apply` command on said YAML files. Once Argo CD is installed, it’s ready to be configured and used.

> There is a project called “[Argo CD Autopilot](https://github.com/argoproj-labs/argocd-autopilot)” upstream, which is used to bootstrap Argo CD onto a cluster. It also takes an opinionated approach to how your git directory structure should look like. This is an open source project created by [Codefresh](codefresh.io/).

Reviewer: https://codefresh.io/
```suggestion
There is a project called “[Argo CD Autopilot](https://github.com/argoproj-labs/argocd-autopilot)” upstream, which is used to bootstrap Argo CD onto a cluster. It also takes an opinionated approach to how your git directory structure should look like. This is an open source project created by [Codefresh](https://codefresh.io/).
```

> # Reconciliation Configurability

Reviewer: @christianh814 et al.

❔ **Do Argo CD and Flux differ at all in the nuances of:**
• Drift Detection and Remediation
• Continuous Operations (CO)
?

❔ To answer that, would you run tests with their latest versions to compare and contrast?

**For example:** If a person runs `kubectl` to manually overwrite Kubernetes cluster's actual state, do Argo CD and Flux _(not only)_ monitor the runtime environment to notice the drift between actual state and desired state, and also converge them to remediate automatically?

My general understanding is that Argo CD and Flux can be used for drift detection and remediation, and Continuous Operations (CO).

**References:**

**[Cloud-native Patterns - Ops Edition - 00:00](https://www.youtube.com/watch?v=fZEFWh4xBEw)**

<img width="1440" alt="Cloud-native Patterns - Ops Edition - 00:00" src="https://user-images.githubusercontent.com/1329685/144722421-5974824e-b70e-485c-abb8-d30fdbfcae97.png">

**[Cloud-native Patterns - Ops Edition - 19:53 - Continuous Operations - Monitor Runtime Env](https://www.youtube.com/watch?v=fZEFWh4xBEw&t=1193s)**

<img width="1440" alt="Cloud-native Patterns - Ops Edition - 19:53 - Continuous Operations - Monitor Runtime Env" src="https://user-images.githubusercontent.com/1329685/144722436-cdaf6d99-bafe-448b-aac7-a94297c8e88d.png">

**[Cloud-native Patterns - Ops Edition - 37:45 - Drift detection and remediation](https://www.youtube.com/watch?v=fZEFWh4xBEw&t=2265s)**

<img width="1440" alt="Cloud-native Patterns - Ops Edition - 37:45 - Drift detection and remediation" src="https://user-images.githubusercontent.com/1329685/144722446-c5d13cb8-79a6-4681-91a8-cc910e181900.png">

**[Cloud-native Patterns - Ops Edition - 42:03 - Continuous Operations - Monitor Runtime Env](https://www.youtube.com/watch?v=fZEFWh4xBEw&t=2523s)**

<img width="1440" alt="Cloud-native Patterns - Ops Edition - 42:03 - Continuous Operations - Monitor Runtime Env" src="https://user-images.githubusercontent.com/1329685/144722458-c60759fa-bb88-4808-b854-5d8bbca8e15a.png">

**Thank you.**

Both projects offer a way to configure how reconciliation works. In this section we’ll focus specifically on the differences between configuring the reconciliation time.

> ## FluxCD

Reviewer: ```suggestion
## Flux CD
```

Flux v2 is built on the [GitOps Toolkit](https://fluxcd.io/docs/components/), which has different components. The Flux reconciliation loop can be configured at each [component](https://fluxcd.io/docs/components/). For example you can have your git repository checked every 5 minutes, but then do the sync every 10 minutes. This gives you the ability to stagger how the reconciliation happens.

## Argo CD

> Argo CD reconciliation is a global setting. This is one setting that can be modified via the [Argo CD configuration ConfigMap](https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/argocd-cm.yaml#L240). By default, this is set to 3 minutes and can be changed by the end user. Although you have a choice of disabling reconciliation globally or per-application, you can only set the reconciliation time globally. This means you cannot set the reconciliation time differently for each [Application](https://argo-cd.readthedocs.io/en/stable/core_concepts/).

Reviewer 4: ```suggestion
Argo CD reconciliation is a global setting. This is one setting that can be modified via the [Argo CD configuration ConfigMap](https://github.com/argoproj/argo-cd/blob/891d9b2f380e1879474ab57c1adf21f5d1678759/docs/operator-manual/argocd-cm.yaml#L246). By default, this is set to 3 minutes and can be changed by the end user. Although you have a choice of disabling reconciliation globally or per-application, you can only set the reconciliation time globally. This means you cannot set the reconciliation time differently for each [Application](https://argo-cd.readthedocs.io/en/stable/core_concepts/).
```

Reviewer: @christianh814 et al.

❔ **Do Argo CD and Flux differ at all in the nuances of:**
• Drift Detection and Remediation
• Continuous Operations (CO)
?

❔ To answer that, would you run tests with their latest versions to compare and contrast?

**For example:** If a person runs `kubectl` to manually overwrite Kubernetes cluster's actual state, do Argo CD and Flux _(not only)_ monitor the runtime environment to notice the drift between actual state and desired state, and also converge them to remediate automatically?

My general understanding is that Argo CD and Flux can be used for drift detection and remediation, and Continuous Operations (CO).

**References:**

**[Cloud-native Patterns - Ops Edition - 00:00](https://www.youtube.com/watch?v=fZEFWh4xBEw)**

<img width="1440" alt="Cloud-native Patterns - Ops Edition - 00:00" src="https://user-images.githubusercontent.com/1329685/144722421-5974824e-b70e-485c-abb8-d30fdbfcae97.png">

**[Cloud-native Patterns - Ops Edition - 19:53 - Continuous Operations - Monitor Runtime Env](https://www.youtube.com/watch?v=fZEFWh4xBEw&t=1193s)**

<img width="1440" alt="Cloud-native Patterns - Ops Edition - 19:53 - Continuous Operations - Monitor Runtime Env" src="https://user-images.githubusercontent.com/1329685/144722436-cdaf6d99-bafe-448b-aac7-a94297c8e88d.png">

**[Cloud-native Patterns - Ops Edition - 37:45 - Drift detection and remediation](https://www.youtube.com/watch?v=fZEFWh4xBEw&t=2265s)**

<img width="1440" alt="Cloud-native Patterns - Ops Edition - 37:45 - Drift detection and remediation" src="https://user-images.githubusercontent.com/1329685/144722446-c5d13cb8-79a6-4681-91a8-cc910e181900.png">

**[Cloud-native Patterns - Ops Edition - 42:03 - Continuous Operations - Monitor Runtime Env](https://www.youtube.com/watch?v=fZEFWh4xBEw&t=2523s)**

<img width="1440" alt="Cloud-native Patterns - Ops Edition - 42:03 - Continuous Operations - Monitor Runtime Env" src="https://user-images.githubusercontent.com/1329685/144722458-c60759fa-bb88-4808-b854-5d8bbca8e15a.png">

**Thank you.**

Reviewer: [Argo CD `timeout.reconciliation`](https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/argocd-cm.yaml#L243:L246)
[#L240](https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/argocd-cm.yaml#L240) → [#L243:L246](https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/argocd-cm.yaml#L243:L246)
```suggestion
Argo CD reconciliation is a global setting. This is one setting that can be modified via the [Argo CD configuration ConfigMap](https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/argocd-cm.yaml#L243:L246). By default, this is set to 3 minutes and can be changed by the end user. Although you have a choice of disabling reconciliation globally or per-application, you can only set the reconciliation time globally. This means you cannot set the reconciliation time differently for each [Application](https://argo-cd.readthedocs.io/en/stable/core_concepts/).
```

Reviewer 3: Argo definitely hand slaps you for using k apply -f manually and sets it back to the last declared state in the Git repo that's sync'ed

# Application Delivery

Both tools support raw Kubernetes manifests (YAML), Kustomize and Helm. Each tool approach is a little different, however. There are other templating mechanisms, like Jsonnet for example. We will focus on the two most popular ones which are Kustomize and Helm.

> ## FluxCD

Reviewer: ```suggestion
## Flux CD
```

<span style="text-decoration:underline;">Raw Kubernetes YAML:</span>

> Flux uses Kustomize heavily. Its[ Kustomize Controller](https://fluxcd.io/docs/components/kustomize/) renders and applies all of the manifests that Flux syncs from the source repository, even when using Raw Kubernetes YAML. Flux will create a `kustommization.yaml` file “on the fly”, if Kustomize wasn’t detected.

Reviewer: kustommization.yaml → kustomization.yaml
```suggestion
Flux uses Kustomize heavily. Its[ Kustomize Controller](https://fluxcd.io/docs/components/kustomize/) renders and applies all of the manifests that Flux syncs from the source repository, even when using Raw Kubernetes YAML. Flux will create a `kustomization.yaml` file “on the fly”, if Kustomize wasn’t detected.
```

> Flux uses Kustomize heavily. Its[ Kustomize Controller](https://fluxcd.io/docs/components/kustomize/) renders and applies all of the manifests that Flux syncs from the source repository, even when using Raw Kubernetes YAML. Flux will create a `kustommization.yaml` file “on the fly”, if Kustomize wasn’t detected.

Reviewer 5: ```suggestion
Flux uses Kustomize heavily. Its[ Kustomize Controller](https://fluxcd.io/docs/components/kustomize/) renders and applies all of the manifests that Flux syncs from the source repository, even when using Raw Kubernetes YAML. Flux will create a `kustomization.yaml` file “on the fly”, if Kustomize wasn’t detected.
```

> When you first bootstrap Flux onto your cluster, it does so by using a Kustomized application called “flux-system” that serves as an entry point for deploying other resources. This entrypoint does a sort of an “autodiscovery” of other YAMLs by doing a Kustomize recursive search in the directory (this is on by default). All someone has to do is create a directory with some YAML, and Flux will “auto detect” it and apply it to the cluster. However, there is no builtin mechanism for ordering. The end user must wait for subsequent reconciliation loops to get a successful application.

Reviewer 4: Flux also supports decrypting secrets, adding merge patches and overriding image names as part of its configuration.

<span style="text-decoration:underline;">Helm:</span>

> Flux also has native support for Helm. Its Helm Controller manages all the Helm charts you want to deploy onto the cluster. Unlike Argo (see below), Flux uses the native Helm golang library to deploy a Helm release onto the cluster. This means that someone can run `helm ls` on the cluster, and it’ll appear as if someone ran `helm install` for the given release. The same issue of “ordering” exists with the Flux CD Helm Controller, however; so a user will still need to use Kustomize, even with their helm charts. You also get the advantage of[ Helm’s Chart Hooks](https://helm.sh/docs/topics/charts_hooks/) for ordering. If you are using a hub/spoke model (more on multi cluster later), you can also see Helm charts installed on other clusters.

Reviewer: `of[` → `of [` 
```suggestion
Flux also has native support for Helm. Its Helm Controller manages all the Helm charts you want to deploy onto the cluster. Unlike Argo (see below), Flux uses the native Helm golang library to deploy a Helm release onto the cluster. This means that someone can run `helm ls` on the cluster, and it’ll appear as if someone ran `helm install` for the given release. The same issue of “ordering” exists with the Flux CD Helm Controller, however; so a user will still need to use Kustomize, even with their helm charts. You also get the advantage of [Helm’s Chart Hooks](https://helm.sh/docs/topics/charts_hooks/) for ordering. If you are using a hub/spoke model (more on multi cluster later), you can also see Helm charts installed on other clusters.
```

<span style="text-decoration:underline;">Things to Note:</span>

Both the Flux Helm Controller and Kustomize Controller CRDs define a _DependsOn_ attribute. That lets you specify dependencies between HelmRelease CRDs, or between Kustomization CRDs, but you can’t currently make a _Kustomization_ depend on a _HelmRelease_. And, although _DependsOn_ enables you to control the load order of collections/groups of YAML files, it does not control the order in which individual YAML files are applied.

## Argo CD

<span style="text-decoration:underline;">Raw Kubernetes YAML:</span>

> Argo CD also supports Kustomize and raw Kubernetes YAML. It can, however, apply raw YAML files directly, without using Kustomize. This can have side effects of Argo CD Applications not syncing right aways as resources need to be reapplied in a retry ([which can be set up per Application](https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/application.yaml#L111-L116)), or use ordering built into Argo CD. Argo CD does the equivalent of `kubectl apply -f` or a `kustomzie build` if Kustomize is detected. This is by design as the Argo CD creators wanted to emulate the `kubectl` client as closely as possible.

Reviewer 4: Does this properly use server side applies, though?

Reviewer 3: Should this fall under "Break Glass" potentially?

> Argo CD also supports Kustomize and raw Kubernetes YAML. It can, however, apply raw YAML files directly, without using Kustomize. This can have side effects of Argo CD Applications not syncing right aways as resources need to be reapplied in a retry ([which can be set up per Application](https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/application.yaml#L111-L116)), or use ordering built into Argo CD. Argo CD does the equivalent of `kubectl apply -f` or a `kustomzie build` if Kustomize is detected. This is by design as the Argo CD creators wanted to emulate the `kubectl` client as closely as possible.

Reviewer 6: You'll probably want to refer to the lines in a specific version/commit. Otherwise the line numbers would change over time.

> Argo CD also supports Kustomize and raw Kubernetes YAML. It can, however, apply raw YAML files directly, without using Kustomize. This can have side effects of Argo CD Applications not syncing right aways as resources need to be reapplied in a retry ([which can be set up per Application](https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/application.yaml#L111-L116)), or use ordering built into Argo CD. Argo CD does the equivalent of `kubectl apply -f` or a `kustomzie build` if Kustomize is detected. This is by design as the Argo CD creators wanted to emulate the `kubectl` client as closely as possible.

Reviewer 6: ```suggestion
Argo CD also supports Kustomize and raw Kubernetes YAML. It can, however, apply raw YAML files directly, without using Kustomize. This can have side effects of Argo CD Applications not syncing right aways as resources need to be reapplied in a retry ([which can be set up per Application](https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/application.yaml#L111-L116)), or use ordering built into Argo CD. Argo CD does the equivalent of `kubectl apply -f` or a `kustomize build` if Kustomize is detected. This is by design as the Argo CD creators wanted to emulate the `kubectl` client as closely as possible.
```

<span style="text-decoration:underline;">Helm:</span>

> Argo CD also has support for Helm, but approaches it in a different way. Argo CD does not use the Helm golang library directly, but instead, uses helm to render the template files and pipes them directly to `kubectl`. Helm chart hooks are converted into Argo CD Syncwaves/Hooks (more on this next) where it can.This has the side effect of users not being able to use the `helm` cli against the cluster because `helm install` wasn’t called in the first place (i.e. `helm ls `will return empty). 

Reviewer 4: > Helm chart hooks are converted into Argo CD Syncwaves/Hooks (more on this next) where it can

This sentence wasn't finished

With Argo CD you have more control over ordering of individual resources (i.e. specific YAML file) with [Syncwaves and Phases](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-waves/). This allows you to apply each YAML manifest one at a time. Using Syncwaves and Phases, Argo CD will wait until one YAML is applied (and healthy) before moving onto applying the next one. 

There is a controller that supplements Argo CD’s Application controller called [Application Sets](https://github.com/argoproj-labs/applicationset). This add-on lets users deploy many Applications across one or many managed clusters.

# Secrets Management

GitOps practice of storing the source of truth in git, has had some contention with respect to storing [Kubernetes Secrets](https://kubernetes.io/docs/concepts/configuration/secret/). This has to do with the fact that Secrets often contain sensitive 	information like passwords or tokens, and these secrets are only encoded.

Both Flux and Argo CD have mechanisms in place to handle the encrypting of secrets.

> ## FluxCD

Reviewer: ```suggestion
## Flux CD
```

> FluxCD provides a guide for managing encrypted secrets with [Mozilla SOPS](https://fluxcd.io/docs/guides/mozilla-sops/) or [Sealed Secrets](https://fluxcd.io/docs/guides/sealed-secrets/) so that Kubernetes secrets can be stored safely in Git, secured by any of OpenPGP, AWS KMS, GCP KMS and Azure Key Vault. Flux’s Kustomization objects natively support decrypting Mozilla SOPS secrets.

Reviewer 4: I'm not sure if this is worth calling out specifically for Flux? The documentation provides guides for these methods, but Flux doesn't provide any additional value on top of these.

Reviewer: ```suggestion
Flux CD provides a guide for managing encrypted secrets with [Mozilla SOPS](https://fluxcd.io/docs/guides/mozilla-sops/) or [Sealed Secrets](https://fluxcd.io/docs/guides/sealed-secrets/) so that Kubernetes secrets can be stored safely in Git, secured by any of OpenPGP, AWS KMS, GCP KMS and Azure Key Vault. Flux’s Kustomization objects natively support decrypting Mozilla SOPS secrets.
```

Reviewer: @RedbackThomson I understand what you mean because both Argo CD and Flux are unopinionated about how secrets are managed.

My hypothesis is that @christianh814 et al. are frequently asked about Secrets Management in GitOps, thus added this `# Secrets Management` to address the FAQ situation.

Reviewer 3: Yes. We are. And we typically point to a part of the CNCF Landscape (sealed secrets, Vault, etc.).

## Argo CD

> Argo CD can also [integrate with a variety of Secret management tools](https://argo-cd.readthedocs.io/en/stable/operator-manual/secret-management/) to store these secrets in Git, safely. It leaves this choice up to the user. Some users have taken advantage of Argo CD’s plugin feature to create plugins, for example, the[ Argo CD Vault Plugin](https://github.com/IBM/argocd-vault-plugin). This “hands off” approach just treats secrets as another _Application_ in the cluster.

Reviewer 6: This is now in argoproj-labs. https://github.com/argoproj-labs/argocd-vault-plugin

# Webhook receivers

Webhooks can be used to help with immediate reconciliation with traditional, event driven, workflows. This doesn’t take the place of the reconciliation loop, but rather acts as an “on demand” reconciliation if you don’t want to wait after making a commit to a repo.

The two tools have similar approaches so we just highlight where to get more information.

> ## FluxCD

Reviewer: ```suggestion
## Flux CD
```

> FluxCD includes support for incoming webhooks that make GitOps pull-based pipelines as fast and responsive as push-based pipelines. [Webhook receivers](https://fluxcd.io/docs/guides/webhook-receivers/) can notify the Flux controllers about changes in Git or Helm repositories, or other sources like Image Repositories.

Reviewer: ```suggestion
Flux CD includes support for incoming webhooks that make GitOps pull-based pipelines as fast and responsive as push-based pipelines. [Webhook receivers](https://fluxcd.io/docs/guides/webhook-receivers/) can notify the Flux controllers about changes in Git or Helm repositories, or other sources like Image Repositories.
```

## Argo CD

Argo CD also has a mechanism to receive [Webhooks](https://argo-cd.readthedocs.io/en/stable/operator-manual/webhook/) to shorten the time it takes to check for updates to a Git repository. Normally this is done every 3 minutes.

> # Alerting and notifications

Reviewer: @christianh814 et al.

**In the context of:**

> (Proposed) GitOps Principle 5. **Operated in a closed loop**
>
>  Software agents observe desired state and meta-state to take actions based on policy when the desired state cannot be reached. This may include things like notifications, rollbacks, etc.

While this draft has [`# Alerting and notifications`](https://github.com/open-gitops/documents/blob/75fe9b46bbc484c58eec3441693185cf586c8717/LANDSCAPE/COMPARISONS/ARGOCD-FLUX/index.md#alerting-and-notifications)

❔ **Do Argo CD and Flux differ at all in the nuances of:**
• rollbacks
• when the desired state cannot be reached
• take actions based on policy
• etc.
?

**References:**
(Proposed) GitOps Principle 5. **Operated in a closed loop**
https://github.com/open-gitops/documents/pull/21#discussion_r709739072
https://github.com/open-gitops/documents/pull/22#discussion_r720456619
https://github.com/open-gitops/documents/issues/24#issuecomment-932754069

**Thank you.**

Reviewer: @christianh814 et al.

**In the context of:**

> (Proposed) GitOps Principle 5. **Operated in a closed loop**
>
>  Software agents observe desired state and meta-state to take actions based on policy when the desired state cannot be reached. This may include things like notifications, rollbacks, etc.

While this draft has [`# Alerting and notifications`](https://github.com/open-gitops/documents/blob/75fe9b46bbc484c58eec3441693185cf586c8717/LANDSCAPE/COMPARISONS/ARGOCD-FLUX/index.md#alerting-and-notifications)

❔ **Do Argo CD and Flux differ at all in the nuances of:**
• rollbacks
• when the desired state cannot be reached
• take actions based on policy
• etc.
?

**References:**
(Proposed) GitOps Principle 5. **Operated in a closed loop**
https://github.com/open-gitops/documents/pull/21#discussion_r709739072
https://github.com/open-gitops/documents/pull/22#discussion_r720456619
https://github.com/open-gitops/documents/issues/24#issuecomment-932754069

**Thank you.**

Reviewer: ```suggestion
# Rollbacks

• Manual `git revert` and `git revert -m` (for merge commits) appear to be the recommended approach to rollback in Git.  Neither Flux nor Argo CD execute the `git` commands on behalf of humans.

## Flux

Flux Helm Controller supports rollback, but it may require manual user intervention to reconcile broken release states.  Flux Kustomize Controller has no mention of rollbacks.

## Argo CD

**Argo CD supports rollback as an emergency measure.  There are issues with pruning and the lack of hooks.

# Alerting and notifications
```
‎

@christianh814 et al.

I searched for rollbacks and here is what I found:

• Manual `git revert` and `git revert -m` (for merge commits) appear to be the recommended approach to rollback in Git.  Neither Flux nor Argo CD execute the `git` commands on behalf of humans.

• **Argo CD** supports rollback as an emergency measure.  There are issues with pruning and the lack of hooks.

• **Flux v2 Helm Controller** supports rollback, but it may require manual user intervention to reconcile broken release states.

• **Flux v2 Kustomize Controller** has no mention of rollbacks.

• **Flux v1 Helm Operator** supports rollbacks, but it may require manual user intervention to recover.

• **Flux v1 (Kustomize) Operator** doesn't support rollbacks.

• **Kubernetes** supports
```kubectl rollout undo deployment/[app]-deployment --to-revision=#```
but it would misalign with GitOps tools.

• **GitOps Working Group Slack** has a conversation about an idea of storing a release artifact from every change applied to production for rollbacks, instead of running `git`.

---

**References below:**

---

**Argo CD**:

From https://argo-cd.readthedocs.io/en/stable/user-guide/commands/argocd_app_rollback/

> `argocd app rollback APPNAME [ID] [flags]`

From https://github.com/argoproj/argo-cd/issues/1927

> When you rollback you cannot:
> 1. Decide to prune resources
> 2. Run hooks

From https://codefresh.io/docs/docs/ci-cd-guides/gitops-deployments/#rolling-back-git-versions

> In the GitOps dashboard you will also see a complete history of all past deployments as recorded in Git. You can select any of the previous versions and rollback your application to the respective version.
>
> ![rollback.png](https://codefresh.io/docs/images/guides/gitops/rollback.png)
> _Rolling back to a previous version_
>
> The Rollback simply informs the cluster to use a different git hash for the sync process. It doesn’t affect your Git repository and ArgoCD will now show your application as out-of-sync (because the last Git commit no longer matches the status of the cluster).
>
> This rollback behavior is best used as an emergency measure after a failed deployment where you want to bring the cluster back to a previous state in a temporary manner. If you wish to keep the current rollback status as a permanent status it is best to use the standard `git reset/revert` commands and change the GitOps repository to its desired state.

---

**Flux v2**:

From https://fluxcd.io/docs/components/helm/helmreleases/

> The remediation action for upgrade failures is by default a rollback

From https://github.com/fluxcd/flux2/issues/1882#issuecomment-932209870

> Flux can correct drift in-cluster, such as a manual helm upgrade that is rollback at the specified interval.

From https://github.com/fluxcd/flux2/issues/1878#issuecomment-962208864

> helmrelease "upgrade retries exhausted" regression
>
> The workaround for now is this:
```flux suspend hr <release_name>```
```flux resume hr <release_name>```
> it will reconcile broken release states such as "exhausted" and "another rollback/release is in progress". 

---

**Flux v1**:

From https://github.com/fluxcd/flux/issues/3201#issuecomment-674854822

> Flux does not rollback nor do we consider adding such a feature as the git state can diverge between sync and apply.

From https://github.com/fluxcd/flux/issues/2132

> Failed Helm releases require manual user intervention to recover
>
> There already is an issue for rollbacks: [#1960](https://github.com/fluxcd/flux/issues/1960) (being worked on in [#2006](https://github.com/fluxcd/flux/pull/2006)), including an explanation on why we can not enable it by default for every HelmRelease.

From https://github.com/fluxcd/flux/issues/1960#issuecomment-488664697

> One of the challenges I am facing is the fact that in case an upgrade fails and it rolls back the release, it will end up in an infinite loop of retrying the upgrade (and rolling back) until a human notices and fixes the error.
>
> This is harmless in theory, it will however result in about 40 nonsense revisions/hour with the default interval settings.

From https://github.com/fluxcd/flux/issues/2221

> **Describe the feature**
>
> A `--previous` flag for `fluxctl release`, which rolls back the given `--workload` to the image before the current image (with filters being applied if configured).
>
> This is useful in more complex CI setups where for example an e2e smoke test can detect (and with this feature easily roll back) a faulty release.
>
> **NB:** for automated resources we would probably need to lock it at the same time, as otherwise it will roll forward again (`--lock`?).
>
> **What would the new user story look like?**
>
> 1. CI performs a release with `fluxctl release --workload <namespace>:<type>/<name>`
> 2. e2e smoke test detects abnormalities, CI now wants to perform a rollback
> 3. CI runs `fluxctl release --workload <namespace>:<type>/<name> --previous`
> 4. resource is now one version behind 'latest'

From https://github.com/fluxcd/flux/issues/2221#issuecomment-626792296

> Maybe this can be solved using kubectl rollout undo, or am I missing something?
>
> UPDATE:
> The problem is that we would break GitOps, right?

---

**Kubernetes**:

From https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#rolling-back-to-a-previous-revision

> **Rolling Back to a Previous Revision**
>
> Now you've decided to undo the current rollout and rollback to the previous revision:
> `kubectl rollout undo deployment/nginx-deployment`
>
> Alternatively, you can rollback to a specific revision by specifying it with --to-revision:
> `kubectl rollout undo deployment/nginx-deployment --to-revision=2`

---

**GitOps Working Group Slack**:

From https://cloud-native.slack.com/archives/C01G9DEE88M/p1637687831177600

> How hard would it be for GitOps controllers/operators/etc. to store a release artifact from every change applied to prod. That way your rollbacks are potentially easier than mucking with git itself. Food for thought.

---

**Git**:

From https://opensource.com/article/18/6/git-reset-revert-rebase-commands

> **Revert or reset?**
>
> Why would you choose to do a revert over a reset operation? If you have already pushed your chain of commits to the remote repository (where others may have pulled your code and started working with it), a revert is a nicer way to cancel out changes for them. This is because the Git workflow works well for picking up additional commits at the end of a branch, but it can be challenging if a set of commits is no longer seen in the chain when someone resets the branch pointer back.
>
> This brings us to one of the fundamental rules when working with Git in this manner: Making these kinds of changes in your local repository to code you haven't pushed yet is fine. But avoid making changes that rewrite history if the commits have already been pushed to the remote repository and others may be working with them.

From https://stackoverflow.com/a/7100005

> **How to revert a merge commit that's already pushed to a remote branch?**
>
> The -m option specifies the parent number. This is because a merge commit has more than one parent, and Git does not know automatically which parent was the mainline, and which parent was the branch you want to un-merge.
>
> In this situation, git revert 8f937c6 -m 1 will get you the tree as it was in 8989ee0, and git revert -m 2 will reinstate the tree as it was in 7c6b236.

---

From https://codefresh.io/about-gitops/pains-gitops-1-0/

> **There is no standard practice for GitOps rollbacks**
>
> The fact that all your cluster history is in Git makes rollbacks in GitOps (supposedly) very easy.
>
> If you want to rollback to a previous version, you simply use a past commit for your sync operation.
>
> In practice, however, it is not entirely clear what exactly we mean by “use a previous commit”. Different people use different ways to rollback:
>
> 1. You can simply point your cluster to a previous Git hash and let your GitOps tool sync that hash. This is the fastest way to rollback, but by definition leaves your cluster in an inconsistent state as the cluster does not contain what is described in the last Git commit.
>
> 2. You can use the standard Git reset, Git revert commands, and again let your GitOps tool perform the sync operation as usual. This keeps the GitOps promise (of having in the cluster what is in the Git repo) but of course, requires manual intervention.
>
> 3. You can have a combination where the GitOps tool itself both syncs a previous Git hash to the cluster and auto-commits (or reverts) to the git repo in order to keep the consistency. This is very complex to accomplish and not all teams want to let their deployment solution have write access to their Git repo.
>
> It goes without saying that different people might want a completely different approach for rollbacks. At the time of writing, however, the present GitOps tools have very little support and guidance on how you perform a rollback in a standard way.

This is another area where both tools are very similar, so we will just focus on the specific implementations.

> ## FluxCD

Reviewer: ```suggestion
## Flux CD
```

> FluxCD controllers emit events when resource statuses change, as new app versions are deployed, and when health checks pass or fail. [Alerting and outgoing notifications](https://fluxcd.io/docs/guides/notifications/) are configured with a Provider resource that connects outgoing webhooks, so Flux can provide status updates through different reporting channels such as Slack or embedded in Git Commit status checks.

Reviewer: ```suggestion
Flux CD controllers emit events when resource statuses change, as new app versions are deployed, and when health checks pass or fail. [Alerting and outgoing notifications](https://fluxcd.io/docs/guides/notifications/) are configured with a Provider resource that connects outgoing webhooks, so Flux can provide status updates through different reporting channels such as Slack or embedded in Git Commit status checks.
```

## Argo CD

Argo CD uses a separate controller for alerting/notifications called [Argo CD Notifications](https://argocd-notifications.readthedocs.io/en/stable/). With Argo CD Notifications, a user can similarly set up providers and outgoing notifications to systems (like Slack or Pagerduty) altering of a specific event that occurred or if a sync failed or not. This project is[ in the process of being moved](https://github.com/argoproj-labs/argocd-notifications#argo-cd-notifications-is-now-part-of-argo-cd) into the main Argo CD codebase and will be part of the “core” Argo CD offering in future releases

# Image updates automation

> Image updating is the process of notifying the GitOps controller that a new image has been created and pushed into a repository. This helps in expediting the process of rolling out new versions of applications.

Reviewer 4: Is it worth calling out that this feature isn't directly tied into GitOps? I think it's great that it's being explored, but it's more related to CD.

Reviewer: Hi @RedbackThomson Cc @christianh814

It is tied into GitOps.

• CI/CD is in [App Delivery TAG's charter](https://github.com/cncf/toc/blob/main/tags/app-delivery.md), and GitOps WG is within App Tag TAG.
• CD is essential in **strong** GitOps which has CD as a foundation, leading to CO (Continuous Operations).

> ## FluxCD

Reviewer: ```suggestion
## Flux CD
```

> FluxCD has a built-in feature called [Image Update Automation](https://fluxcd.io/docs/guides/image-update/), which can be configured to scan for new container images in image repositories that are used by applications in the cluster. FluxCD directly enters the new version into an existing YAML manifest in the form of a git commit.

Reviewer: ```suggestion
Flux CD has a built-in feature called [Image Update Automation](https://fluxcd.io/docs/guides/image-update/), which can be configured to scan for new container images in image repositories that are used by applications in the cluster. Flux CD directly enters the new version into an existing YAML manifest in the form of a git commit.
```

## Argo CD

> Argo CD uses a separate controller called the ["Argo CD Image Updater"](https://argocd-image-updater.readthedocs.io/en/stable/), which takes a similar approach of “writing back” image updates to the YAML manifest in the git repository. The image updater will do this in the form of a[ Kustomize image update](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/image.md). Currently, the tool will only work if you’re using Kustomize or Helm, not not raw (non-kustomized) YAML files.

Reviewer 4: ```suggestion
Argo CD uses a separate controller called the ["Argo CD Image Updater"](https://argocd-image-updater.readthedocs.io/en/stable/), which takes a similar approach of “writing back” image updates to the YAML manifest in the git repository. The image updater will do this in the form of a[ Kustomize image update](https://github.com/kubernetes-sigs/kustomize/blob/master/examples/image.md). Currently, the tool will only work if you’re using Kustomize or Helm and not raw (non-kustomized) YAML files.
```

# Web UI

A Web UI doesn’t change the functionality of either tool, but it’s worth noticing the differences as Flux CD doesn’t have a UI and Argo CD does. Here we just called out the differences and notable information.

> ## FluxCD

Reviewer: ```suggestion
## Flux CD
```

Flux CD, by itself, does not have a Web UI. There is an [experimental UI ](https://github.com/fluxcd/webui)that was started about a year ago, but nothing official. There are opensource platforms, like [Weave GitOps Core](https://github.com/weaveworks/weave-gitops) that are based on Flux and have a UI. But the upstream Flux project currently does not have a UI.

## Argo CD

> Argo CD has a rich UI and is very UI centric. Argo CD was built with the UI in mind, so it’s been there from the beginning. It also integrates with Argo CD’s ACL Policies (more on that next) and has the ability to show a “read only” dashboard. This is included “out of the box” with Argo CD, or it can be disabled by installing [Argo CD Core](https://argo-cd.readthedocs.io/en/stable/operator-manual/installation/#core).

Reviewer: @christianh814 Since # Web UI exist, was there ever a # CLI in this document?

# RBAC

> With RBAC, each tool not only takes a different technical approach, but also a different philosophical approach. \

Reviewer 4: ```suggestion
With RBAC, each tool not only takes a different technical approach, but also a different philosophical approach.
```

> ## FluxCD

Reviewer: ```suggestion
## Flux CD
```

Flux relies strictly on the RBAC capabilities of Kubernetes, and does not manage its own users, permissions, identity, authentication or authorization. There are no “flux users” or “flux admins”. Flux’s permissions are limited by the Role bound to the ServiceAccount it runs as. This is by design. Flux tries to conform to Kubernetes practices of least privilege for the Service Account.

## Argo CD

> Argo CD has an[ ACL based RBAC](https://argo-cd.readthedocs.io/en/stable/operator-manual/rbac/), independent of Kubernetes. Argo CD also has a separate group/user mechanism, which is also independent of Kuberenetes. Normally, a user sets up Argo CD’s **_ServiceAccount_** with full admin, and manages users using Argo CD’s user/group mechanism. Or a user can use a combination of the two to get the desired outcome.

Reviewer 4: ```suggestion
Argo CD has an [ACL based RBAC](https://argo-cd.readthedocs.io/en/stable/operator-manual/rbac/), independent of Kubernetes. Argo CD also has a separate group/user mechanism, which is also independent of Kubernetes. Normally, a user sets up Argo CD’s **_ServiceAccount_** with full admin, and manages users using Argo CD’s user/group mechanism. Or a user can use a combination of the two to get the desired outcome.
```

The RBAC/Access system on Argo CD is very granular. Although it can be complex to set up, it is very flexible. The Argo CD authentication mechanism can be integrated with[ any provider that supports OIDC](https://argo-cd-docs.readthedocs.io/en/latest/operator-manual/sso/) or brokers it using Dex.

# Multi Cluster

Multi-cluster management is a challenge. Good news; both tools can help with multi-cluster management and deployment of your workloads to different clusters you may have. Both tools have similar models but different implementations.

> ## FluxCD

Reviewer: ```suggestion
## Flux CD
```

> FluxCD has the concept of connecting to the remote clusters via Kustomization’s (and HelmRelease’s) _KubeConfig_ setting. FluxCD uses a “pull-based” design, where there needs to be an instance of FluxCD running inside of each cluster. A centralized “management cluster” is another possible configuration, where one Flux instance manages workloads on many clusters. “Multi Cluster '' is achieved through [design and process](https://github.com/fluxcd/flux2-multi-tenancy), via a reference architecture. [Other tools, like CAPI, need to be brought in](https://www.youtube.com/watch?v=7W27tAv7Tvs), to install and manage the clusters.

Reviewer 4: ```suggestion
FluxCD has the concept of connecting to the remote clusters via Kustomization’s (and HelmRelease’s) _KubeConfig_ setting. FluxCD uses a “pull-based” design, where there needs to be an instance of FluxCD running inside of each cluster. A centralized “management cluster” is another possible configuration, where one Flux instance manages workloads on many clusters. “Multi Cluster" is achieved through [design and process](https://github.com/fluxcd/flux2-multi-tenancy), via a reference architecture. [Other tools, like CAPI, need to be brought in](https://www.youtube.com/watch?v=7W27tAv7Tvs), to install and manage the clusters.
```

Reviewer: ```suggestion
Flux CD has the concept of connecting to the remote clusters via Kustomization’s (and HelmRelease’s) _KubeConfig_ setting. Flux CD uses a “pull-based” design, where there needs to be an instance of Flux CD running inside of each cluster. A centralized “management cluster” is another possible configuration, where one Flux instance manages workloads on many clusters. “Multi Cluster '' is achieved through [design and process](https://github.com/fluxcd/flux2-multi-tenancy), via a reference architecture. [Other tools, like CAPI, need to be brought in](https://www.youtube.com/watch?v=7W27tAv7Tvs), to install and manage the clusters.
```

Reviewer: Use smart quotes “” for opening and closing quotes
```suggestion
FluxCD has the concept of connecting to the remote clusters via Kustomization’s (and HelmRelease’s) _KubeConfig_ setting. FluxCD uses a “pull-based” design, where there needs to be an instance of FluxCD running inside of each cluster. A centralized “management cluster” is another possible configuration, where one Flux instance manages workloads on many clusters. “Multi Cluster” is achieved through [design and process](https://github.com/fluxcd/flux2-multi-tenancy), via a reference architecture. [Other tools, like CAPI, need to be brought in](https://www.youtube.com/watch?v=7W27tAv7Tvs), to install and manage the clusters.
```

![flux-mc](IMAGES/flux-mc.png)

> FluxCD supports multi-cluster management with a single Flux instance by setting [KubeConfig](https://fluxcd.io/docs/components/kustomize/kustomization/#specification) setting. FluxCD also allows the "push" behaviour in this sense by connecting and deploying YAMLS to target clusters via KubeConfig.

Reviewer: ```suggestion
Flux CD supports multi-cluster management with a single Flux instance by setting [KubeConfig](https://fluxcd.io/docs/components/kustomize/kustomization/#specification) setting. Flux CD also allows the "push" behaviour in this sense by connecting and deploying YAMLS to target clusters via KubeConfig.
```
